contracts/strategies/HStrategy.sol

```diff
@@ -0,0 +1,851 @@
+        );
+        oracleParams = newOracleParams;
+        emit UpdateOracleParams(tx.origin, msg.sender, newOracleParams);
+    }
+
+    /// @notice updates parameters of the capital ratios and deviation. Can be called only by admin
+    /// @param newRatioParams the new parameters
+    function updateRatioParams(RatioParams calldata newRatioParams) external {
+        _requireAdmin();
+        require(
+            newRatioParams.erc20CapitalRatioD <= DENOMINATOR &&
+                newRatioParams.minCaptialDeviationD <= DENOMINATOR &&
+                newRatioParams.erc20CapitalRatioD >= newRatioParams.minCaptialDeviationD &&
+                newRatioParams.minRebalanceDeviationD > 0 &&
+                newRatioParams.minRebalanceDeviationD <= DENOMINATOR,
+            ExceptionsLibrary.INVARIANT
+        );
+        ratioParams = newRatioParams;
+        emit UpdateRatioParams(tx.origin, msg.sender, newRatioParams);
+    }
+
+    /// @notice manual pulling tokens from vault. Can be called only by admin
+    /// @param fromVault the address of the vault to pull tokens from
+    /// @param toVault the address of the vault to pull tokens to
+    /// @param tokenAmounts the amount of tokens to be pulled
+    /// @param vaultOptions additional options for `pull` method
+    function manualPull(
+        IIntegrationVault fromVault,
+        IIntegrationVault toVault,
+        uint256[] memory tokenAmounts,
+        bytes memory vaultOptions
+    ) external {
+        _requireAdmin();
+        fromVault.pull(address(toVault), tokens, tokenAmounts, vaultOptions);
+    }
+
+    /// @notice rebalance method. Need to be called if the new position is needed
+    /// @param restrictions the restrictions of the amount of tokens to be transferred
+    /// @param moneyVaultOptions additional parameters for pulling for `pull` method for money vault
+    /// @return actualPulledAmounts actual transferred amounts
+    function rebalance(RebalanceTokenAmounts memory restrictions, bytes memory moneyVaultOptions)
+        external
+        returns (RebalanceTokenAmounts memory actualPulledAmounts, uint256[] memory burnedAmounts)
+    {
+        _requireAtLeastOperator();
+        IUniswapV3Pool pool_ = pool;
+        (, int24 tick, , , , , ) = pool_.slot0();
+        _hStrategyHelper.checkSpotTickDeviationFromAverage(tick, address(pool_), oracleParams, _uniV3Helper);
+        burnedAmounts = _partialRebalanceOfUniV3Position(restrictions, tick);
+        actualPulledAmounts = _capitalRebalance(restrictions, moneyVaultOptions, tick);
+    }
+
+    /// @notice rebalance, that if needed burns old univ3 position and mints new
+    /// @param restrictions the restrictions of the amount of tokens to be transferred
+    /// @param tick current price tick
+    /// @return burnedAmounts actual transferred amounts of tokens from position while burn
+    function _partialRebalanceOfUniV3Position(RebalanceTokenAmounts memory restrictions, int24 tick)
```

Comment:
maybe somehow add an opportunity to rebalance immediately if we changed intervalWidth but tick is yet inside the old position?